routing/hopiterator.go

```diff
@@ -0,0 +1,204 @@
+package routing
+
+import (
+	"bytes"
+	"encoding/hex"
+
+	"io"
+
+	"github.com/btcsuite/golangcrypto/ripemd160"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lightning-onion"
+	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcutil"
+)
+
+// HopID represents the id which is used by propagation subsystem in order to
+// identify lightning network node.
+type HopID [ripemd160.Size]byte
+
+// NewHopID creates new instance of hop form node public key.
+func NewHopID(pubKey []byte) HopID {
+	var routeId HopID
+	copy(routeId[:], btcutil.Hash160(pubKey))
+	return routeId
+}
+
+// String returns string representation of hop id.
+func (h HopID) String() string {
+	return hex.EncodeToString(h[:])
+}
+
+// IsEqual checks does the two hop ids are equal.
+func (h HopID) IsEqual(h2 HopID) bool {
+	return bytes.Equal(h[:], h2[:])
+}
+
+// HopIterator interface represent the entity which is able to give route
+// hops one by one. This interface is used to have an abstraction over the
+// algorithm which we use to determine the next hope in htlc route.
+// Currently we have two implementation if ths interface is:and also
```

Comment:
The comments below this line (what's currently implemented), can be safely omitted.